### PR TITLE
feat(curriculum): explain shorthand stat calculator

### LINKED ISSUE #55689: stats calculator seed code changes between step 33 and 34 without explanation
### Describe the Issue

The js beta project to build a statistics calculator has an unexpected change in the editor's code between step 33 and step 34. No explanation is given for this change.
In step 33 we are initially shown a forEach statement which looks like this:
```
  array.forEach(el => {

  })
```
We are asked to fill it in. Since back in step 29 we were shown sample code to help us write the new forEach, we probably end up writing it in a very similar way to the sample code there: https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-advanced-array-methods-by-building-a-statistics-calculator/step-29

Then in step 34, the forEach suddenly becomes this one liner:
`array.forEach(el => counts[el] = (counts[el] || 0) + 1)`

But it is not acknowledged at all that the code changed. We don't even notice it till step 35 when it returns back to the main editing area.

I've seen people on the forum copy this code verbatim for other projects. However, I don't believe they know what it is doing (other than it counts the number of occurrences that is). Most people still struggle with truthies at this point of their learning so if we really want to write the code like this (and if we do, what was the point of step 29?), it would be great if there was a step between 33 and 34 that explains this shorthand.

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-advanced-array-methods-by-building-a-statistics-calculator/step-34

### Your code

N/A

### Expected behavior

code campers write should not change so fundamentally without explanation in the project

### Screenshots

_No response_

### System

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by a2937**
Checklist:

- [X] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [X] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

Closes #55689

I added an extra step and did my best to explain how to transform the arrow function into the shorthand seen in the later steps.

### CODE REVIEW

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+An example of such a expression is the following: 
+
+```js
+const numbersArr = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4];
+const counts = {};
+numbersArr.forEach(el => counts[el] = (counts[el] || 0) + 1)
+```
+
+In this example, the `numbersArr` has each element looped through like before. Only this time, it removes the 
+if statement and replaces it with an `or` operator. This `or` operator says if `counts[el]` is *falsy* to return the other value.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md:21]:
```suggestion
In this example, the `numbersArr` has each element looped through like before. Only this time, it removes the if statement and replaces it with an `or` operator. This `or` operator says if `counts[el]` is falsy to return the other value. 
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+```js
+const numbersArr = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4];
+const counts = {};
+numbersArr.forEach(el => counts[el] = (counts[el] || 0) + 1)
+```
+
+In this example, the `numbersArr` has each element looped through like before. Only this time, it removes the 
+if statement and replaces it with an `or` operator. This `or` operator says if `counts[el]` is *falsy* to return the other value. 
+
+
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md:23]:
```suggestion
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+An value can be considered *falsy* if it is `null`, `undefined`, or otherwise can be evaluated as false.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md:24]:
```suggestion
An value can be considered <dfn>falsy</dfn> if it is `null`, `undefined`, or otherwise can be evaluated as false. 
```

Though I do feel like we've taught this before.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+title: Step 34
+challengeType: 0
+dashedName: step-34
+---
+
+# --description--
+
+While `getMode` is looking great, the `forEach` loop can simplified further. In order to do so, the function will be turned into an `arrow function expression`. This kind of expression is known for not using curly braces and shorter syntax. 
+
+An example of such a expression is the following:
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md:12]:
```suggestion
An example of such an expression is the following: 
```

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
I'm not sure about the wording, either. Will probably need to give this another pass.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md (hunk 3 above)
Comment (a2937) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md:24]: We might have explained it in the shopping cart curriculum though it might not hurt to explain it multiple times.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+const numbersArr = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4];
+const counts = {};
+numbersArr.forEach(el => counts[el] = (counts[el] || 0) + 1)
+```
+
+In this example, the `numbersArr` has each element looped through like before. Only this time, it removes the if statement and replaces it with an `or` operator. This `or` operator says if `counts[el]` is falsy to return the other value. 
+
+An value can be considered <dfn>falsy</dfn> if it is `null`, `undefined`, or otherwise can be evaluated as false. 
+
+Either way the value of `counts[el]` gets incremented by one. Turn the existing `forEach` loop into an `arrow function expression`.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md:24]:
```suggestion
Your `forEach` loop is fully functional, but it's very verbose. You can actually simplify it further.

To do so, you will need to use an <dfn>arrow function expression</dfn>. This kind of function expression does not use curly braces, opting for a more compact single-line syntax.

Here is an example of an arrow function expression: 

```js
const numbersArr = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4];
const counts = {};
numbersArr.forEach(el => counts[el] = (counts[el] || 0) + 1)
```

In this example, `forEach` still iterates through every value in the array. The callback is now an arrow function expression.

The `(counts[el] || 0)` falls back to `0` when `counts[el]` is `undefined`, or another falsy value. This ensures that `counts[el]` will be incremented by 1, or initialised to 1 if the property does not have a truthy value.

Replace the callback in your `forEach` to an arrow function expression like the example.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+# --description--
+
+While `getMode` is looking great, the `forEach` loop can simplified further. In order to do so, the function will be turned into an `arrow function expression`. This kind of expression is known for not using curly braces and shorter syntax. 
+
+An example of such an expression is the following: 
+
+```js
+const numbersArr = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4];
+const counts = {};
+numbersArr.forEach(el => counts[el] = (counts[el] || 0) + 1)
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md:17]: Not sure how I feel about an example that could pretty much be copy-pasted.

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md (hunk 6 above)
Comment (a2937) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md:17]: Oh. Yeah that's an issue. :( I should probably fix it.

**Comment by lasjorg** [comment]
The expression body is unrelated to the shorter syntax.

`forEach` always returns `undefined` and it is meant for side effects. It doesn't matter if the callback has an implicit or explicit return, the callback can't return anything.

In fact, we might want to avoid the expression body syntax with `forEach`. Seeing as the callback can't return anything, implying otherwise by using an implicit return could be seen as dishonest syntax. Don't get me wrong, I have certainly written `forEach` loops with an expression body before. I'm just not sure if we should be promoting it.

```
numbersArr.forEach((el) => {
  counts[el] = (counts[el] || 0) + 1;
});
```


Also, this is just my personal opinion, but the way it is worded makes it sound like they should prefer the much less readable version because the readable version is too verbose. Furthermore, simplification is not the same as shorter syntax, in this case, I would say the shorter version is not a simplification because it is harder to read.

All this just to say, I think we have to be careful about what we promote and how we do it.

**Comment by ojeytonwilliams** [comment]
Hi @a2937 could you take a look at these conflicts?

@naomi-lgbt it's ready for another review.

**Comment by naomi-lgbt** [comment]
> it's ready for another review.

Negative - we're still using an example that passes the tests.

**Comment by a2937** [comment]
I apologize for forgetting about this PR. I will go ahead and update the tests and resolve the merge conflicts.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md (hunk 6 above)
Comment (a2937) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md]: I've updated the example.

**Comment by a2937** [comment]
There were other challenges that changed the meta. The best way to resolve the issue was to rebase this PR. Hope that is understandable.

**Comment by ojeytonwilliams** [comment]
Hi @a2937, the issue here is just that the braces are dropped without explanation. This goes against the general pattern that the solution to the current challenge should be the seed for the next one.

Also, both syntaxes `() => 1` and `() => { return 1; }` have been introduced before. At most, we need to remind people of the concise form - we don't need to define it.

I think it's good to add an extra step, so we can ask for the shorter syntax and justify it. I just wouldn't define it again.

**Comment by a2937** [comment]
Okay so don't define arrow syntax again. Got it.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66f59e667892883f47abe6d5.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+```js
+const expected = {'2': 1, '4': 2, '5': 1};
+assert.deepEqual(getMode([4, 4, 2, 5]), expected)
+```
+
+Your function should use an `arrow function expression`. 
+
+```js
+assert.match(getMode.toString(), /counts\[el\]\s*=\s*\(counts\[el\]\s*\|\|\s*0\)\s*\+\s*1/);
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66f59e667892883f47abe6d5.md:35]: Just a thought... We already test the function implementation by asserting the returned value is correct. So rather than a regular expression against the function body, should we regex this bit:

```
array.forEach(el =>
```

And assert it is not followed by a `{`?

**Review by naomi-lgbt — APPROVED** [review]
Approving this as it's perfectly fine as is. Leaving a non-blocking idea.

**Comment by ojeytonwilliams** [comment]
Sorry, @a2937, I've been going back through the curriculum and I can't find anywhere that introduces the form without curly braces. We just sort of start using it without telling anyone, which isn't great.

This makes me think this isn't really a problem with the stat calculator. Perhaps it would be better to add a lesson introducing them to https://www.freecodecamp.dev/learn/javascript-algorithms-and-data-structures-v8/#learn-basic-string-and-array-methods-by-building-a-music-player once we've introduced arrow functions in general. i.e. after step 11. Then all the stat calculator lesson needs to mention is "hey, we can clean this function up using the concise syntax".

What do you reckon?

Also, @naomi-lgbt any thoughts?

**Comment by naomi-lgbt** [comment]
It is taught in step 15: https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-basic-string-and-array-methods-by-building-a-music-player/step-15

**Comment by ojeytonwilliams** [comment]
Ah, so it is, my mistake! Thanks for finding that.

In which case, all we need to do is mention that they should refactor to use the implicit return.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66f59e667892883f47abe6d5.md (hunk 7 above)
Comment (lasjorg) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66f59e667892883f47abe6d5.md:35]: That code also isn't the only valid expression. You could use a ternary.

So yes, we shouldn't force that single expression as the only valid solution.

**Comment by lasjorg** [comment]
If we absolutely want to keep the seed code, we should create a step that shows and explains it as part of teaching how to convert the callback block body to an expression body. We could show how to do it using a ternary as well. Then just have them write it, even though we gave them the solution as part of the explanation. But we should still make sure any valid solution using an expression body will pass.

Giving them the solution is not much different from how we teach in step 30-34 anyway. It is entirely possible to learn from code without coming up with the solution yourself. Sure, this might not be in line with how we run most of the curriculum, but I think using that format occasionally isn't all bad.

---

Super quick example of what I mean (wip and all).

````md
There is another way to write the `forEach`. Instead of using a block body `() => {}` for the callback, you can use an expression body `() =>`.
Without a code block inside the callback, we have to convert the statements we had into an expression. *Maybe explain why*

For example, we could write it like so:
```
array.forEach(el => counts[el] = (counts[el] || 0) + 1)
```

*Code explanation if needed*

Or we could use a ternary:

```
array.forEach(el => counts[el] = counts[el] ? counts[el] + 1 : 1);
```

*Code explanation if needed*

Convert the `forEach` callback from using a block body to an expression body.
````

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66f59e667892883f47abe6d5.md (hunk 7 above)
Comment (a2937) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66f59e667892883f47abe6d5.md:35]: That'd probably be pretty reasonable.

**Comment by a2937** [comment]
I don't wanna give campers a copy pastable solution to the challenge as an example. Not really. It's bad enough about Step 30 and the following four steps. 

The point of this was to have a step that bridges the gap between the two different pieces of seed code. However as you said, the seed code solution is not the only answer.

**Comment by lasjorg** [comment]
> The point of this was to have a step that bridges the gap between the two different pieces of seed code.

But it doesn't bridge the gap. Requiring an expression body doesn't solve anything related to the issue.

We can't "coerce" the camper into writing, and more importantly understanding, our seed code just by forcing them to use an expression body for the callback.

What we could do is teach how to convert the callback to use an expression body and convert the statements to an expression. And in doing so, also explain our seed code in the process. We can change the object name and values so all that is left is the "shape" of the code. We are teaching something new, by explaining a refactor.

Our seed code is not the logical conclusion that anyone will reach just by us requiring an expression body. Doing so will not explain the seed code, nor will it particularly help guide the camper to write it just like we have it in the seed code.

A lot of people on the forum can barely understand the seed code. Or have it wrong, but think they understand it. For example, some people think the initial value assignment is `0` and not `0 + 1` and I don't blame them, because the code is not that easy to read.

TL;DR We have to explain the seed code, or remove it.

**Comment by a2937** [comment]
I really don't wanna keep the seed code at the end of the day. It's a little too complicated for the average person. Unfortunately the best option I can really think of is to change the seed code so the sudden switch doesn't happen.

**Comment by lasjorg** [comment]
We may have to guide them a bit more than just asking for a ternary. A lot of beginners will put the assignments inside the ternary conditions and while that works it isn't what we expect and not the proper use of a ternary.

For example, they might write:
```
array.forEach(el => counts[el] ? counts[el] += 1 : counts[el] = 1);
```

---
Here is another version.

````md
There is another way to write the `forEach`. Instead of using a block body `() => {}` for the callback, you can use an expression body `() =>`.

You will have to convert the `if...else` statements into an expression. Write the expression as a ternary and use a single assignment for the ternary.

```
assigment = condition ? exprIfTrue : exprIfFalse
```

Convert the `forEach` callback to use an expression body and replace the statements with a ternary.
````

**Comment by lasjorg** [comment]
I think we just add the assignment part to the ternary code examples.

```js
assigment = condition ? exprIfTrue : exprIfFalse
```

**Review by majestic-owl448 — APPROVED** [review]
